### ISSUE TITLE
Code for DPO-finetuning?

### ISSUE THREAD

**Issue by KasenYoung** [KwaiVGI/VideoAlign#4]
Thanks for your great work. I wonder if you can release the code for finetuning video models using your reward model. This procedure can show your work's strengths better.

**Comment by yifan123** [comment]
Thank you for your kind words and interest in our work!

Regarding your question: The DPO implementation in VideoAlign was built on an internal codebase for training closesourced models, which unfortunately isn’t publicly available at this time. For replication, if you have a repository for fine-tuning a flow-based video generation model, modifying it to incorporate flow-DPO is relatively easy:

1. **Base Model**: Start with open-source flow-matching video models (e.g., [Wan2.1-T2V-1.3B](https://huggingface.co/Wan-AI/Wan2.1-T2V-1.3B-Diffusers)).

2. **Data Prep**: Structure preference data using paired (win/lose) videos. Follow the common DPO practice: `win_videos = batch[:N]`, `lose_videos = batch[N:]` (see [DiffusionDPO data handling](https://github.com/SalesforceAIResearch/DiffusionDPO/blob/main/train.py#L843)).

3. **Loss Implementation**: Adapt the flow-DPO loss from VideoAlign's Appendix E pseudocode. You can also refer to the implementation of [DiffusionDPO's loss logic](https://github.com/SalesforceAIResearch/DiffusionDPO/blob/main/train.py#L1125).

**Key Insight**: For existing LoRA-tuned video models (flow-matching based), adding Flow-DPO mainly requires modifying the data pipeline and loss computation.

**Comment by ArlenCHEN** [comment]
hi @yifan123,

Thanks for those very useful recommendations. 

I just wonder if you have ever encountered the issue of loss quickly reaching 0. Does this mean a reward hacking? Any insight is appreciated!!

Thanks!

**Comment by yifan123** [comment]
It seems that the difference (w_diff - l_diff) is very large, causing the loss to quickly reach 0. This usually happens when the pattern between the win and lose pairs you constructed is particularly obvious, making it easy for the model to distinguish between the two types of data. The most straightforward solution is to increase the beta, possibly by 100 times its current value, to prevent the model from overfitting. You could also modify your data, as win and lose pairs that are too easy to distinguish indicate that the lose data is not a hard negative sample.

![Image](https://github.com/user-attachments/assets/2ae501a2-63b6-4b61-9181-cb2c72edbc23)

**Comment by ArlenCHEN** [comment]
@yifan123 thanks for your quick reply. what you said really makes a lot of sense. seems making proper pairs is a so much important and not easy topic for video dpo.

do you think other methods, for example, some reference-free methods, like this one: SimPO: Simple Preference Optimization
with a Reference-Free Reward, can solve/mitigate the  problem?

**Comment by ArlenCHEN** [comment]
another thing: i encountered the issue when i used a full-parameters tuning. a lora-based training is still on the way. 

do you think using full-param instead of lora can be one of the reasons?

**Comment by yifan123** [comment]
Thanks for your interesting for our work!

As discussed in the paragraph Discussion on $\beta_t$ in Section 4.1, our experiments show that setting $\beta_t = \beta(1 - t)^2$ degrades performance, whereas using a constant $\beta_t$ yields better results.

**Comment by caihaunqai** [comment]
> Thank you for your kind words and interest in our work!
> 
> Regarding your question: The DPO implementation in VideoAlign was built on an internal codebase for training closesourced models, which unfortunately isn’t publicly available at this time. For replication, if you have a repository for fine-tuning a flow-based video generation model, modifying it to incorporate flow-DPO is relatively easy:
> 
> 1. **Base Model**: Start with open-source flow-matching video models (e.g., [Wan2.1-T2V-1.3B](https://huggingface.co/Wan-AI/Wan2.1-T2V-1.3B-Diffusers)).
> 2. **Data Prep**: Structure preference data using paired (win/lose) videos. Follow the common DPO practice: `win_videos = batch[:N]`, `lose_videos = batch[N:]` (see [DiffusionDPO data handling](https://github.com/SalesforceAIResearch/DiffusionDPO/blob/main/train.py#L843)).
> 3. **Loss Implementation**: Adapt the flow-DPO loss from VideoAlign's Appendix E pseudocode. You can also refer to the implementation of [DiffusionDPO's loss logic](https://github.com/SalesforceAIResearch/DiffusionDPO/blob/main/train.py#L1125).
> 
> **Key Insight**: For existing LoRA-tuned video models (flow-matching based), adding Flow-DPO mainly requires modifying the data pipeline and loss computation.

Hello， I follow your paper to achieve Flow-DPO. However, the loss is quite large when beta is bigger, such as 1.5, when beta is 160. However, I trained for 2000 iterations, but the performance does not show a large difference between the DPO model and the SFT model, and is even the same in most prompts. Can you give me some advice?

**Comment by yifan123** [comment]
How did you construct the positive and negative pairs? My suggestion is to collect *N* images/videos using the same prompt, then select the one with the highest reward as the positive and the one with the lowest reward as the negative. This way, DPO will significantly outperform SFT, and the effect becomes more pronounced as *N* increases. Generally, I recommend using *N* ≥ 16.

**Comment by caihaunqai** [comment]
Is the learning rate, time shift important for the training? If it does, which do you use?

**Comment by yifan123** [comment]
The time shift is likely not critical, but I used the same shift as in pretraining. The learning rate can be found in Table 10 of the paper.


![Image](https://github.com/user-attachments/assets/d92b31c1-11d5-4079-9e1d-a8f4465578c6)

**Comment by yifan123** [comment]
> Thanks for your great work. I wonder if you can release the code for finetuning video models using your reward model. This procedure can show your work's strengths better.

For all, thank you so much for your interest! I've provided a script for applying Flow-DPO to text-to-image tasks:
[https://github.com/yifan123/flow\_grpo/blob/main/scripts/train\_sd3\_dpo.py](https://github.com/yifan123/flow_grpo/blob/main/scripts/train_sd3_dpo.py)

**Comment by mariolarcher** [comment]
> > Thanks for your great work. I wonder if you can release the code for finetuning video models using your reward model. This procedure can show your work's strengths better.
> 
> For all, thank you so much for your interest! I've provided a script for applying Flow-DPO to text-to-image tasks: https://github.com/yifan123/flow_grpo/blob/main/scripts/train_sd3_dpo.py

Thanks @yifan123 !
